group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -761,6 +776,58 @@ public ClassicGroup classicGroup(
         }
     }
 
+    /**
+     * Validates the online upgrade if the Classic Group receives a ConsumerGroupHeartbeat request.
+     *
+     * @param classicGroup A ClassicGroup.
+     * @return the boolean indicating whether it's valid to online upgrade the classic group.
+     */
+    private boolean validateOnlineUpgrade(ClassicGroup classicGroup) {
+        if (!consumerGroupMigrationPolicy.isUpgradeEnabled()) {
+            log.debug("Online upgrade is invalid because the consumer group {} migration config is {} so online upgrade is not enabled.",
+                classicGroup.groupId(), consumerGroupMigrationPolicy);
+            return false;
+        } else if (classicGroup.isInState(DEAD)) {
+            log.debug("Online upgrade is invalid because the classic group {} is in DEAD state.", classicGroup.groupId());
+            return false;
+        } else if (!classicGroup.usesConsumerGroupProtocol()) {
+            log.debug("Online upgrade is invalid because the classic group {} has protocol type {} and doesn't use the consumer group protocol.",
```

Comment:
nit: `Cannot upgrade classic group {} to consumer group because the group does not use the consumer embedded protocol.`